### PR TITLE
Collections expressions: LDM update

### PR DESCRIPTION
**Pull request by cston**
Update on collection expressions for LDM 2023-08-14.

### CODE REVIEW

File: meetings/working-groups/collection-literals/CL-LDM-2023-08-14.md [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+static List<T[]> AsListOfArray<T>(List<T[]> arg) => arg;
+```
+
+The [*type inference*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#1163-type-inference) change is handled with a rule for *input type inference* and a corresponding rule for *output type inference*: see [proposal](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#type-inference).
+
+*Input type inference* is necessary for walking into the element expressions to infer from explicitly-typed lambda parameters, from tuple element types, or from nested collection expressions.
+```c#
+InputTypeInference([(int y) => { }]); // InputTypeInference<int>()
+
+static void InputTypeInference<T>(List<Action<T>> x) { }
```
Comment (CyrusNajmabadi) [inline meetings/working-groups/collection-literals/CL-LDM-2023-08-14.md:78]: woudl this work with a different delegate tyep from Func/Action?

**Review by CyrusNajmabadi — APPROVED** [review]

File: meetings/working-groups/collection-literals/CL-LDM-2023-08-14.md [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+*An array or concrete type is already preferred over an interface if the array or concrete type implements that interface. And if the array or concrete type does not implement the interface, we shouldn't prefer one over the other. Instead, the overload resolution proposal should be: Prefer a span over an array or interface if the iteration type of the span is implicitly convertible to the iteration type of the array or interface.*
+
+## Type inference
+
+*See proposal: [Type inference](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#type-inference)*
+
+Method type inference involving collection expressions allows inferring the [*iteration type*](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/statements.md#1295-the-foreach-statement) of a collection type. There is no support for inferring the containing *collection type*.
+
+The *iteration type* is defined by `foreach`. If a *collection initializer type* implements `IEnumerable` only, the iteration type is `object`.
```
Comment (jnm2) [inline meetings/working-groups/collection-literals/CL-LDM-2023-08-14.md:59]: Being super nitpicky, a counterexample is where the type implements `IEnumerable` explicitly and also has a GetEnumerator method (which doesn't implement any interface method) which returns a different enumerator type with a different `foreach` element type.

File: meetings/working-groups/collection-literals/CL-LDM-2023-08-14.md [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+## Conversions
+
+*See proposal: [Conversions](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#conversions)*
+
+There is a *collection expression conversion* from a collection expression to the following types. These represent the valid *target types* for a collection expression.
+
+* single-dimensional arrays
+* span types
+* types with a [*builder method*](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#create-methods)
+* generic interfaces implemented by `List<T>`
```
Comment (jnm2) [inline meetings/working-groups/collection-literals/CL-LDM-2023-08-14.md:97]: What if `List<T>` adds a new generic interface implementation?

File: meetings/working-groups/collection-literals/CL-LDM-2023-08-14.md [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+```
+
+For all but collection initializer types, there must be an *implicit conversion* to `T` for each *expression element* in the collection expression, and an *implicit conversion* to `T` for the *iteration type* of each *spread element*.
+```c#
+string[] a = [x, y, z];
+int[] b = [1, null];           // error: cannot convert 'null' to 'int'
+ImmutableArray<int> c = [..a]; // error: cannot convert 'string' to 'int'
+```
+
+For collection initializer types, there must be an applicable instance or extension `Add` method for each *expression element* in the collection expression, and an applicable `Add` method for an argument of the *iteration type* of each *spread element*
```
Comment (jnm2) [inline meetings/working-groups/collection-literals/CL-LDM-2023-08-14.md:115]:
```suggestion
For collection initializer types, there must be an applicable instance or extension `Add` method for each *expression element* in the collection expression, and an applicable `Add` method for an argument of the *iteration type* of each *spread element*.
```

File: meetings/working-groups/collection-literals/CL-LDM-2023-08-14.md [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+*See proposal: [Ref safety](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-12.0/collection-expressions.md#ref-safety)*
+
+For collection expressions of `ref struct` target types, the compiler *may* allocate the storage for the collection on the callstack when the following hold:
+* The collection has a *known length* and is not empty
+* The runtime supports *inline array* types
+* The collection expression has *local scope*
+
+An empty collection requires no allocation for storage.
+
+A `ReadOnlySpan<T>`, where `T` is one of several *primitive types*, and that contains constant values only is stored in the assembly data section and does not allocate at the use-site.
```
Comment (jnm2) [inline meetings/working-groups/collection-literals/CL-LDM-2023-08-14.md:146]:
```suggestion
A `ReadOnlySpan<T>`, where `T` is one of several *primitive types* and where the collection expression contains constant values only, is stored in the assembly data section and does not allocate at the use-site.
```

**Review by jnm2 — APPROVED** [review]

File: meetings/working-groups/collection-literals/CL-LDM-2023-08-14.md (hunk 1 above)
Comment (cston) [reply meetings/working-groups/collection-literals/CL-LDM-2023-08-14.md:78]: This doesn't depend on `Action<T>`. A delegate such as `delegate void D<T>(T t);` would work too.

File: meetings/working-groups/collection-literals/CL-LDM-2023-08-14.md (hunk 3 above)
Comment (cston) [reply meetings/working-groups/collection-literals/CL-LDM-2023-08-14.md:97]: Added a question, thanks.

**Pull request merged**